crypto/evp/p_lib.c

```diff
@@ -1386,17 +1386,19 @@ static int evp_pkey_reset_unlocked(EVP_PKEY *pk)
     if (pk == NULL)
         return 0;
 
-    memset(pk, 0, sizeof(*pk));
+    if (pk->lock) {
+      const size_t offset = (unsigned char *)&pk->lock - (unsigned char *)pk;
+      memset(pk, 0, offset);
+      memset(&pk->lock + 1, 0, sizeof(*pk) - offset - sizeof(pk->lock));
```

Comment:
This is correct in terms of address computation. I am not sure it is strictly correct in regards to C standard. Isn't effectively dereferencing array (of a single element) past its bounds an undefined operation? But is memset() actually counted as dereference?

Not sure if `(unsigned char *)&pk->lock + sizeof(pk->lock) would make it any different though.
What about `memset((unsigned char *)pk + offset + sizeof(pk->lock), 0, sizeof(*pk) - offset - sizeof(pk->lock));` ?

Reply:
That worked like a charm, thanks again.

Reply:
I have a memory of the C standard allowing arrays to be indexed one position past their end without it becoming an undefined operation.

The alternative is better.

Reply:
It is defined behavior to compute the value of a pointer that points one (but not two!) past the end of an object/array, yes.
Most of the text about (un)defined behavior with respect to this sort of access is written in terms of "object"s, and in this case one would say that the `evp_pkey_st` is the object, so I'm not really convinved there would be a problem in the old formulation either.

Another thing to note, though, is that the compiler is not exactly obligated to lay things out in memory in the order that the fields are declared in the structure definition.  In practice it mostly is, due to the requirement to be able to link with other objects, the strictures of platform ABI specs, etc., though there are some "hardening" compiler options that will randomize struct layouts.  It looks like we're currently only preserving the lock, not the refcount, but if we were trying to skip the memset for both fields we'd want to add some sort of check/assertion that the fields are laid out in the order we expect.